### DISCUSSION TITLE
通过readelf -sW查不到对应符号为什么也能hook成功呢，有没有大佬帮忙解释下

### DISCUSSION THREAD

**Discussion by threeloe** [bytedance/android-inline-hook#63] [Q&A]
以_ZN7android10uirenderer12renderthread11RenderProxy10setStoppedEb为例在libhwui.so中，能在Android 8.1上通过readelf查看到，但在Andorid 12上就查不到，这也就是说不在. dynsym和.symtab中吧，为什么还能hook到呢

**Comment by Colibrow** [comment]
* `shadowhook_dlsym_dynsym` 只能查找 `.dynsym` 中的符号，速度较快。
* `shadowhook_dlsym_symtab` 能查找 `.symtab` 和 `.symtab in .gnu_debugdata` 中的符号，但是速度较慢。
* `shadowhook_dlsym` 会先尝试在 `.dynsym` 中查找符号，如果找不到，会继续尝试在 `.symtab` 和 `.symtab in .gnu_debugdata` 中查找。

**Reply by threeloe** [reply]
求问 .symtab in .gnu_debugdata 的符号怎么查看，搜了一圈没太懂 xz+dd+readelf 具体是怎么操作的

**Reply by Colibrow** [reply]
1. llvm-readelf -S libhwui.so
`
[27] .gnu_debugdata    PROGBITS        0000000000000000 830ae3 04b7f0 00      0   0  1
`
2. dd if=libhwui.so of=debugdata.xz bs=1 skip=0x830ae3 count=0x04b7f0 
3. xz -d debugdata.xz
4. llvm-readelf -s debugdata

**Reply by threeloe** [reply]
Thank you,bro
